akka-distributed-data/src/main/scala/akka/cluster/ddata/DurableStore.scala

```diff
@@ -0,0 +1,245 @@
+ * `failureMsg` to the `replyTo`.
+ */
+object DurableStore {
+  final case class Store(key: String, data: ReplicatedData, reply: Option[StoreReply])
+  final case class StoreReply(successMsg: Any, failureMsg: Any, replyTo: ActorRef)
+
+  case object Load
+  final case class LoadData(data: Map[String, ReplicatedData])
+  case object LoadCompleted
+}
+
+object MapDbDurableStore {
+  def props(config: Config): Props =
+    Props(new MapDbDurableStore(config))
+
+  private case object Commit extends DeadLetterSuppression
+}
+
+final class MapDbDurableStore(config: Config) extends Actor with ActorLogging {
+  import DurableStore._
+  import MapDbDurableStore.Commit
+
+  val serializer = SerializationExtension(context.system).serializerFor(classOf[DataEnvelope])
+  val commitInterval = config.getString("mapdb.commit-interval").toLowerCase match {
+    case "off" ⇒ Duration.Zero
+    case _     ⇒ config.getDuration("mapdb.commit-interval", MILLISECONDS).millis
+  }
+
+  val db = {
+    val file = config.getString("mapdb.file") match {
+      case path if path.endsWith(".ddata") ⇒
+        new File(path)
+      case dir ⇒
+        new File(dir, s"${context.system.name}-${self.path.parent.name}-${Cluster(context.system).selfAddress.port.get}.ddata")
+    }
+    file.getParentFile.mkdirs()
+
+    val builder = DBMaker.fileDB(file)
+      .fileMmapEnableIfSupported()
+      .fileMmapPreclearDisable()
+      .allocateStartSize(config.getBytes("mapdb.allocate-start-size"))
+      .allocateIncrement(config.getBytes("mapdb.allocate-increment"))
+      .concurrencyDisable() // only access from this actor
+      .closeOnJvmShutdown()
+
+    if (config.getBoolean("mapdb.write-ahead-log"))
+      builder.transactionEnable() // WAL for crash safety
+
+    builder.make()
+  }
+  val store: HTreeMap[String, ReplicatedData] =
+    db.hashMap("store")
+      .keySerializer(Serializer.STRING)
+      .valueSerializer(new MapdbReplicatedDataSerializer(context.system))
+      .createOrOpen()
+
+  val uncommitted = new java.util.HashMap[String, ReplicatedData]
+
+  override def postStop(): Unit = {
+    super.postStop()
+    commit()
+    db.close()
+  }
+
+  def receive = init
+
+  def init: Receive = {
+    case Load ⇒
+      try {
+        val loadData = LoadData(store.entrySet.asScala.map { entry ⇒
+          entry.getKey → entry.getValue
+        }(breakOut))
+        if (loadData.data.nonEmpty)
+          sender() ! loadData
+        sender() ! LoadCompleted
+      } catch {
+        case NonFatal(e) ⇒
+          log.error(e, "failed to load data")
+          sender() ! LoadCompleted
+      } finally
+        context.become(active)
+  }
+
+  def active: Receive = {
+    case Store(key, data, reply) ⇒
+      try {
+        if (commitInterval.length == 0) {
+          // optimize calls from put, no need to deserialize old value, using ThreadLocal
+          MapdbReplicatedDataSerializer.callFromPut.set(MapdbReplicatedDataSerializer.NotUsedOldValue)
```

Comment:
> A. Why is it needed at all? (MapDB deficiency?)
Yes, it's an inefficiency: https://github.com/jankotek/mapdb/issues/769

> B. Why a threadlocal instead of a threadsafe toggle on the serializer instance (I would have found that much easier to follow)?

That would mean that the serializer would not be thread-safe, which would be fine for the usage in this actor here, but would not if someone later refactor this to be used outside of an actor.

>C. Could we have a small abstraction for put that does this instead of repeating the Not-used-put-finally-clear?

sure, I can add method with a call by name block, at the cost of allocation